src/Http/Routing/src/Matching/HostMatcherPolicy.cs

```diff
@@ -0,0 +1,330 @@
+                    // So, loop and add it to all states.
+                    foreach (var kvp in edges)
+                    {
+                        kvp.Value.Add(endpoint);
+                    }
+                }
+                else
+                {
+                    // OK this endpoint matches specific hosts
+                    foreach (var kvp in edges)
+                    {
+                        // The edgeKey maps to a possible request header value
+                        var edgeKey = kvp.Key;
+
+                        for (var j = 0; j < endpointKeys.Length; j++)
+                        {
+                            var endpointKey = endpointKeys[j];
+
+                            if (edgeKey.Equals(endpointKey))
+                            {
+                                kvp.Value.Add(endpoint);
+                                break;
+                            }
+                            else if (edgeKey.HasHostWildcard && endpointKey.HasHostWildcard &&
+                                edgeKey.Port == endpointKey.Port && edgeKey.MatchHost(endpointKey.Host))
+                            {
+                                kvp.Value.Add(endpoint);
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return edges
+                .Select(kvp => new PolicyNodeEdge(kvp.Key, kvp.Value))
+                .ToArray();
+        }
+
+        public PolicyJumpTable BuildJumpTable(int exitDestination, IReadOnlyList<PolicyJumpTableEdge> edges)
+        {
+            if (edges == null)
+            {
+                throw new ArgumentNullException(nameof(edges));
+            }
+
+            // Since our 'edges' can have wildcards, we do a sort based on how wildcard-ey they
+            // are then then execute them in linear order.
+            var ordered = edges
+                .Select(e => (host: (EdgeKey)e.State, destination: e.Destination))
+                .OrderBy(e => GetScore(e.host))
+                .ToArray();
+
+            return new HostPolicyJumpTable(exitDestination, ordered);
+        }
+
+        private int GetScore(in EdgeKey key)
+        {
+            // Higher score == lower priority.
+            if (key.MatchesHost && !key.HasHostWildcard && key.MatchesPort)
+            {
+                return 1; // Has host AND port
+            }
+            else if (key.MatchesHost && !key.HasHostWildcard)
+            {
+                return 2; // Has host
+            }
+            else if (key.MatchesHost && key.MatchesPort)
+            {
+                return 3; // Has wildcard host AND port
+            }
+            else if (key.MatchesHost)
+            {
+                return 4; // Has wildcard host
+            }
+            else if (key.MatchesPort)
+            {
+                return 5; // Has port
```

Comment:
What does this look like? `:8080`? How is it different from `*:8080`?

Reply:
They're the same and both would hit this point.

Host wildcard means `*.contoso.com`. `*` means no host.